rfcs/20190308-addons-proxy-maintainership.md

```diff
@@ -0,0 +1,174 @@
+granted after substantial contribution has been made in order to limit the 
+number of users with write perission. Contributions can come in the form of 
+issue closings, bug fixes, new code, or optimizing existing code. Submodule 
+maintainership can be granted with a lower barrier for entry as this will not 
+include write permissions to the repo.
+
+Additionally, Proxy maintainers are encouraged to co-maintain existing TensorFlow 
+Addons subpackages/submodules. If a potential maintainer chooses to do so, 
+there are a few things to consider first:
+
+* If the subpackage has an existing maintainer, the potential maintainer 
+must communicate with them first. The existing maintainer would need to 
+approve the co-maintainer role, and might give the potential maintainer 
+a specific task to complete before granting co-maintainership.
+
+* If the subpackage has major bugs, the potential maintainer 
+can provide fixes and expedite their inclusion as a listed
+proxy maintainer.
+
+### Privileges and Responsibitilies of Proxied Maintainers
+
+Subpackage maintainers will be granted write access to the repo and will be 
+designated in the [CODEOWNERS](https://help.github.com/en/articles/about-code-owners) 
+file.
+
+### Resolving Bugs
+
+Proxy maintainers are expected to handle bugs against the subpackages 
+they own. This includes resolving those issues once their fix is 
+merged by a proxy maintainer.
+
+If a proxy maintainer fails to provide sufficient support for their
+subpackage they will be designated as inactive. This status will be taken into 
+account during the periodic review of their subpackage/submodule.
+
+### Retiring from Proxy Maintainership
+
+If a proxy maintainer decides that they no longer wish to maintain one 
+or more of their subpackages, they **must** commit to the following procedure:
+
+* Send an email to addons@tensorflow.org and list all code that they are 
+no longer able to maintain. It is usually a good idea to shortly describe the 
+state of code, e.g. whether the packages have open bugs, whether they are difficult 
+to maintain, etc.
+
+* The proxy maintainer must remove themselves from the `CODEOWNERS.md` 
+file, with `<!--maintainer-needed-->` listed in the place of 
+maintainers. This will help other developers to find subpackages with no 
+maintainers.
+
+## Repository Growth and Review 
+### Periodic Review
+
+Given the nature of this repository, subpackages and submodules may become less 
+and less useful to the community as time goes on. In order to keep the 
+repository sustainable, we'll be performing bi-annual reviews of our code to 
+ensure everything still belongs within the repo. Contributing factors to this 
+review will be:
+
+1. Number of active maintainers
+2. Amount of OSS use
+3. Amount of issues or bugs attributed to the code
+
+Because we encourage TF-Addons to be used in production, we'll plan on a 
+graceful deprecation process, instead of a swift removal.
```

Comment:
similar here, should we define what is `graceful` in this case. Could be too early, but potentially very valuable promises for people introducing addons in production

Reply:
Yeah, this is probably better done sooner rather than later. I do think we'll reserve deprecating things unless they're truly unused in the community at least for a while.

Reply:
explicit is better than implicit ;)

Reply:
Agreed that "explicit is better than implicit". 

IMHO, "deprecation" can come from different reasons, eg if the lib is short of maintenance/full of bug, or a better solution is available to solve the same problem, which makes the current API less useful.

I think API can be categorized into 3 groups: Suggested/Discouraged/Deprecated.
Suggested: Well maintained API and encourage user to use.
Discouraged: Better alternative is available, the API is kept for historical reason; or just in the waiting period to be deprecated; or the API is short of maintenance, which might cause issue in prod.
Deprecated: Use at your own risk, subject to be deleted after certain peroid. 

The status change between these three is: Suggested <-> Discouraged -> Deprecated

The period between an API being marked as deprecated and being deleted can be 90 days.
1. If addons does release monthly, there will be 2-3 releases before an API is deleted. The release notes could give user enough warning.
2. It is not too long comparing to 180 days, nor too short as 30 days, so that people can fix their code.
3. Google project zero uses 90 days window for security fix, which is a trade off between risk and time to respond. For more comparison of different time window, please see https://googleprojectzero.blogspot.com/search?q=90+days.

Also we should be mindful the amount of Discouraged API. They should either be fixed, or move to deprecated. Keeping Discouraged API will make long tail and cause the repo hard to maintain.

Reply:
I think this is a solid plan... would you like to submit a modification to this PR? Not sure if you have access to the  branch, but @dynamicwebpaige should be able to help.

Reply:
I think its fine for either you or Paige to just update the PR.

Reply:
Updating now! And very much like the plan for Suggested -> Discouraged -> Deprecated. :+1: